uefi/src/proto/shell/mod.rs

```diff
@@ -0,0 +1,239 @@
+//! EFI Shell Protocol v2.2
+
+use core::{ffi::c_void, mem::MaybeUninit, ptr::NonNull};
+
+use uefi_macros::{unsafe_guid, Protocol};
+
+use crate::{CStr16, Event, Handle, Result, Status};
+
+use super::media::file::FileInfo;
+
+/// TODO
+#[repr(C)]
+#[unsafe_guid("6302d008-7f9b-4f30-87ac-60c9fef5da4e")]
+#[derive(Protocol)]
+pub struct Shell {
+    execute: extern "efiapi" fn(
+        parent_image_handle: *const Handle,
+        commandline: *const CStr16,
+        environment: *const *const CStr16,
+        out_status: *mut Status,
+    ) -> Status,
+    get_env: usize,
+    set_env: usize,
+    get_alias: usize,
+    set_alias: usize,
+    get_help_text: usize,
+    get_device_path_from_map: usize,
+    get_map_from_device_path: usize,
+    get_device_path_from_file_path: usize,
+    get_file_path_from_device_path: usize,
+    set_map: usize,
+
+    get_cur_dir: usize,
+    set_cur_dir: usize,
+    open_file_list: usize,
+    free_file_list: usize,
+    remove_dup_in_file_list: usize,
+
+    batch_is_active: extern "efiapi" fn() -> bool,
+    is_root_shell: usize,
+    enable_page_break: extern "efiapi" fn(),
+    disable_page_break: extern "efiapi" fn(),
+    get_page_break: usize,
+    get_device_name: usize,
+
+    get_file_info: usize,
+    set_file_info: usize,
+    open_file_by_name: usize,
+    close_file: extern "efiapi" fn(file_handle: ShellFileHandle) -> Status,
+    create_file: extern "efiapi" fn(
+        file_name: &CStr16,
+        file_attribs: u64,
+        out_file_handle: *mut ShellFileHandle,
+    ) -> Status,
+    read_file: usize,
+    write_file: usize,
+    delete_file: extern "efiapi" fn(file_handle: ShellFileHandle) -> Status,
+    delete_file_by_name: extern "efiapi" fn(file_name: &CStr16) -> Status,
+    get_file_position: usize,
+    set_file_position: usize,
+    flush_file: usize,
+    find_files: extern "efiapi" fn(
+        file_pattern: *const CStr16,
+        out_file_list: *mut *mut ShellFileInfo,
+    ) -> Status,
+    find_files_in_dir: usize,
+    get_file_size: usize,
+
+    open_root: usize,
+    open_root_by_handle: usize,
+
+    execution_break: Event,
+
+    major_version: u32,
+    minor_version: u32,
+    register_guid_name: usize,
+    get_guid_name: usize,
+    get_guid_from_name: usize,
+    get_env_ex: usize,
+}
+
+impl Shell {
+    /// TODO
+    pub fn execute(
+        &self,
+        parent_image: Handle,
+        command_line: &CStr16,
+        environment: &[&CStr16],
+    ) -> Result<Status> {
+        // This is required because the compiler won't turn `&[&Cstr16]`
+        // or `*const &Cstr16 into *const *const CStr16`
+        let mut out_status: MaybeUninit<Status> = MaybeUninit::uninit();
```

Comment:
Is `MaybeUninit` the right abstraction? At this point, we know that the memory is initialized, as `&[&CStr16]` can't be uninitialized?

I think, `let environment = environment as *const *const CStr16;` should be sufficient. 

Thoughts?

Reply:
That comment is in the wrong spot and causing a misunderstanding, I think. The MaybeUninit is an out-parameter for the return code of the executed script and not related to environment.

Reply:
> That comment is in the wrong spot and causing a misunderstanding, I think.

Yikes, sorry! Yes, you're right.

Reply:
No worries, the comment placement was my fault anyway :)

As well, `let environment = environment as *const *const CStr16;` seems like it should be a valid cast but surprisingly isn't according to rustc, so I had to do it in two parts.